### DISCUSSION TITLE
Sensor for the charging costs (in eur/kWh)

### DISCUSSION THREAD

**Discussion by rrozema** [hif2k1/battery_sim#135] [Ideas]
I would like to keep track of the price (in eur/kWh) of the energy stored in my (simulated) batteries. This would f.e. be extremely useful for decision making on when to charge or discharge the battery for optimal monetary results: if the current price for electricity is <= the price I paid for charging the battery, it makes sense to pause the battery. But when the current price for electricity is > the price of the energy in the battery, the battery should discharge as much the usage of the home. Any variations with bandwidths, incl. and excl. VAT etc. should of course be possible too.

Does anyone have a suggestion on how to keep track of only the price paid for charging the battery, diluting that price with the weighted price for the amount charged? I don't think the existing price sensors can be used for this purpose, or can they?

My suggested sensor for the price of the stored energy would not adjust its value when the battery discharges. When the battery discharges only the SOC (in kWh) lowers in value,  the price sensor (in eur/kWh) would remain the same. When the battery charges however, both the SOC (in kWh) and the price (in eur/kWh) may change: the SOC sensor will get the amount charged added, the price sensor will have some value added that takes the current price for electricity and the amount charged. I can not get my head around the correct formula so that when an amount of C(1) kWh gets added at a current price of P(1) eur/kWh to the battery that has a previous charge of C(0) at a price of P(0). For simplicity I chose to leave the losses due to inefficiencies while charging and discharging out of it, but these of course need to be included at some point.

**Comment by rrozema** [comment]
I think I've got the formula to calculate the new price after some x amount of energy was charged:
NewPrice = ((SOC * OldPrice) + (max(Charged, 0) * (1/efficiency) * GridPrice)) / (SOC + max(Charged, 0) )

NewPrice = the new price of the energy in the battery (EUR/kWh)
SOC = State of Charge before charge (kWh)
OldPrice = the old price of the energy in the battery (EUR/kWh)
Charged = amount of energy added to the battery (kWh) MUST be positive only!
Efficiency = charger efficiency (unit less, f.e. 0.8)
GridPrice = current price for energy consumed (EUR/kWh)
max( x, y) = returns the highest of the both parameters x and y.

@hif2k1 : would it be possible to create a helper component that implements the above logic and can be added both to your simulated batteries and actual batteries? i.e. so that we can use it to implement charging/discharging logic based on the actual price paid for the contents of a battery?

**Comment by rrozema** [comment]
@hif2k1 do you have any comments, please? I think the above would be very helpful in implementing a working strategy aimed at return of investment in combination with a dynamic contract.

**Comment by rrozema** [comment]
I've created this triggered template sensor that pretty much does what I want it to do, although I think it doesn't properly reload yet upon a restart of HA. Any suggestions?

```
template:
  - trigger:
      - trigger: state
        entity_id:
          - sensor.battery_sim_diy_100k
        not_from:
          - "unavailable"
          - "unknown"
        not_to:
          - "unavailable"
          - "unknown"
    sensor:
      - name: "battery_sim: diy 100k charging costs"
        unique_id: "f997992b-ddf1-44ed-8004-edffa5e68219"
        state: |-
          {% set current_costs = states("sensor.electricity_meter_energy_consumption_price") | float %}
          {% set efficiency = state_attr("sensor.battery_sim_diy_100k", "efficiency") | float(1) %}
          {% set soc_old = trigger.from_state.state | float %}
          {% set soc_new = trigger.to_state.state | float %}
          {% set soc_delta = (soc_new - soc_old) | round(5) %}
          {% set price_old = this.state | float(current_costs) %}
          {{ 0.0 if soc_new <= 0.0 
            else (price_old if soc_delta <= 0.0
                else (((soc_old * price_old) + (soc_delta * (1/effciency) * current_costs)) / soc_new) | round(10)) }}
        device_class: monetary
        state_class: measurement
        unit_of_measurement: "EUR/kWh"
```

**Comment by rrozema** [comment]
I can now easily compare the value of this sensor to the (hourly changing) price for electricity to see if I should switch on or off the 'discharge only' switch on my simulated batteries. I don't have the logic in place yet to automatically switch the battery mode, but I can already make a nice graph, demonstrating the effect of the battery's inefficiency on my discharging strategy. Both are in EUR/kWh, so this makes the graph easy too:
![image](https://github.com/user-attachments/assets/fd041f64-5b9b-408d-a352-487b8c8fca9d)
The yellow line is the charging costs: when this is over the current price, it makes no sense to use the energy from the battery, as using energy from the grid is cheaper. When the yellow line is over the blue line I should switch the battery off. When the yellow line is under the blue line, I switch on the 'discharge only'. I do not even want to store my excess solar at that time: with my dynamic contract with Tibber gives solar energy going back into the grid me a higher profit than storing it in the battery because of the battery's inefficiencies. I only want to charge the battery when the price is as low as possible: which keeps the yellow line low and thus the profit high.

**Comment by dewi-ny-je** [comment]
Interesting idea. I'll look into it.

**Comment by dewi-ny-je** [comment]
It's now done in my testing fork, which will be integrated into this repository after sufficient testing is performed.

For the time being you can test the feature from
https://github.com/dewi-ny-je/battery_sim/releases/tag/v2.4.0

Keep in mind that this fork is based on the latest v2.3.0 pre-release, not on 1.3.x, which already has many changes implemented, some breaking ones
https://github.com/hif2k1/battery_sim/releases/tag/v2.3.0

If you have issues, open a ticket in that repository.
